Add support for Get/Post/Delete operations in API Shield Endpoint Management

djhworld opened the pull request:
This change adds support for the following API Shield related endpoints related to endpoint management:

- [Add operations to a zone](https://developers.cloudflare.com/api/operations/api-shield-endpoint-management-add-operations-to-a-zone)
- [Delete an operation](https://developers.cloudflare.com/api/operations/api-shield-endpoint-management-delete-an-operation)
- [Retrieve information about an operation](https://developers.cloudflare.com/api/operations/api-shield-endpoint-management-retrieve-information-about-an-operation)


## Description

This change adds support for API Shield related endpoints related to managing operations in API Shield Endpoint Management. This is so they can be used from the library and will enable this feature to be added to terraform in the future. 

## Has your change been tested?

- Unit tests
- Local testing using API


## Types of changes

What sort of change does your code introduce/modify?

- [x] New feature (non-breaking change which adds functionality)


## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] This change is using publicly documented in [cloudflare/api-schemas](https://github.com/cloudflare/api-schemas) 
      and relies on stable APIs.

Supersedes #1054

djhworld committed:
APISHI-2353 Add support for Get/Post/Delete operations in API Shield Endpoint Management

jacobbednarz committed:
Merge branch 'master' into dharper/APISHI-2353

maciej reviewed (commented):
lgtm

jacobbednarz commented on api_shield_operations.go:
```diff
@@ -0,0 +1,221 @@
+
+import (
+	"context"
+	"encoding/json"
```
we are intentionally using `github.com/goccy/go-json` (perf, handling some edge cases better) instead of the stdlib json library so we should update that here too.

djhworld replied:
fixed in https://github.com/cloudflare/cloudflare-go/pull/1397/commits/583a0d227aeb5d06bbfbb9d9df4aedf0eb887219

jacobbednarz commented on api_shield_operations.go:
```diff
@@ -0,0 +1,221 @@
+)
+
+// APIShieldCreateOperation should be used when creating an operation in API Shield Endpoint Management.
+type APIShieldCreateOperation struct {
```
check out the conventions for naming CRUD operation params - https://github.com/cloudflare/cloudflare-go/blob/master/docs/conventions.md#methods.

djhworld replied:
fixed in https://github.com/cloudflare/cloudflare-go/pull/1397/commits/2c4380c0cb9ca9ec185937eafaae318b542cfe3d

djhworld replied:
note this struct is used as part of `CreateAPIShieldOperationsParams` - it's basically what the user is expected to set when creating an operation, the `CreateAPIShieldOperationsParams` accepts a slice of these. 

I've update the name of this to `APIShieldBasicOperation` to (hopefully?) make this clearer

jacobbednarz commented on api_shield_operations.go:
```diff
@@ -0,0 +1,221 @@
+// GetAPIShieldOperation returns information about an operation
+//
+// API documentation https://developers.cloudflare.com/api/operations/api-shield-endpoint-management-retrieve-information-about-an-operation
+func (api *API) GetAPIShieldOperation(ctx context.Context, rc *ResourceContainer, operationID string, params *APIShieldGetOperationParams) (*APIShieldOperation, error) {
```
the CRUD methods should only ever take a maximum of 3 parameters (see https://github.com/cloudflare/cloudflare-go/blob/master/docs/conventions.md#methods)

djhworld replied:
fixed in https://github.com/cloudflare/cloudflare-go/pull/1397/commits/2c4380c0cb9ca9ec185937eafaae318b542cfe3d

jacobbednarz commented on api_shield_operations.go:
```diff
@@ -0,0 +1,221 @@
+func (api *API) GetAPIShieldOperations(ctx context.Context, rc *ResourceContainer, params *APIShieldGetOperationsParams) ([]APIShieldOperation, ResultInfo, error) {
+	uri := fmt.Sprintf("/zones/%s/api_gateway/operations", rc.Identifier)
+	if params != nil {
+		uri = strings.Join([]string{uri, params.Encode()}, "?")
```
instead of manually building the URL, you can use [`buildURI`](https://github.com/cloudflare/cloudflare-go/blob/f968533e2861e3d84c52e6e5f194c31038604580/utils.go#L13) and `url` struct tags to handle this automatically (and consistently). 

example: https://github.com/cloudflare/cloudflare-go/blob/f968533e2861e3d84c52e6e5f194c31038604580/zone_hold.go#L30

djhworld replied:
fixed in https://github.com/cloudflare/cloudflare-go/pull/1397/commits/2c4380c0cb9ca9ec185937eafaae318b542cfe3d

jacobbednarz commented on api_shield_operations.go:
```diff
@@ -0,0 +1,221 @@
+// GetAPIShieldOperations retrieve information about all operations on a zone
+//
+// API documentation https://developers.cloudflare.com/api/operations/api-shield-endpoint-management-retrieve-information-about-all-operations-on-a-zone
+func (api *API) GetAPIShieldOperations(ctx context.Context, rc *ResourceContainer, params *APIShieldGetOperationsParams) ([]APIShieldOperation, ResultInfo, error) {
```
generally, `Get*` methods should only get a single resource. i think you want `List*` here instead.

djhworld replied:
fixed in https://github.com/cloudflare/cloudflare-go/pull/1397/commits/2c4380c0cb9ca9ec185937eafaae318b542cfe3d

jacobbednarz commented on api_shield_operations.go:
```diff
@@ -0,0 +1,221 @@
+// PostAPIShieldOperations add one or more operations to a zone.
+//
+// API documentation https://developers.cloudflare.com/api/operations/api-shield-endpoint-management-add-operations-to-a-zone
+func (api *API) PostAPIShieldOperations(ctx context.Context, rc *ResourceContainer, operations []APIShieldCreateOperation) ([]APIShieldOperation, error) {
```
we generally favour `Create*`/`Update*` over `Post*` to better demonstrate the intent and separation of concerns if needed down the track.

djhworld replied:
fixed in https://github.com/cloudflare/cloudflare-go/pull/1397/commits/2c4380c0cb9ca9ec185937eafaae318b542cfe3d

jacobbednarz commented on api_shield_operations.go:
```diff
@@ -0,0 +1,221 @@
+
+// Encode is a custom method for encoding APIShieldGetOperationsParams into a usable HTTP
+// query parameter string.
+func (a APIShieldGetOperationsParams) Encode() string {
```
this method isn't required if you swap to the `buildURI` helper.

djhworld replied:
fixed in https://github.com/cloudflare/cloudflare-go/pull/1397/commits/2c4380c0cb9ca9ec185937eafaae318b542cfe3d

jacobbednarz commented on api_shield_operations.go:
```diff
@@ -0,0 +1,221 @@
+
+// Encode is a custom method for encoding APIShieldGetOperationParams into a usable HTTP
+// query parameter string.
+func (a APIShieldGetOperationParams) Encode() string {
```
this method isn't required if you swap to the `buildURI` helper.

djhworld replied:
fixed in https://github.com/cloudflare/cloudflare-go/pull/1397/commits/2c4380c0cb9ca9ec185937eafaae318b542cfe3d

jacobbednarz commented:
solid first pass -- only some conventions/consistency based updates here to get it inline with the existing library.

thanks!

hc2116 reviewed (commented):
The logic itself looks good to me, cannot comment on the integration of this to cloudflare-go

djhworld committed:
Update Get/Post/Delete operations in API Shield Endpoint Management to follow library conventions

- Using `github.com/goccy/go-json` over `encoding/json`
- Renamed `*GetOperations` to `*ListOperations`
- Changed methods to accept 3 parameters only, and used method specific
  `*Params` structs
- Renamed `*Param` structs to be paired with the methods they are
  associated with (e.g. `GetAPIShieldOperationParam`)
- Use `buildURI` method + added struct tags so query parameters are
  encoded using go-queryparams (removed custom `Encode` methods)

djhworld commented:
@jacobbednarz have resolved the comments, fixes are in this commit https://github.com/cloudflare/cloudflare-go/pull/1397/commits/2c4380c0cb9ca9ec185937eafaae318b542cfe3d

jacobbednarz committed:
swap LastUpdated to *time.Time

jacobbednarz commented:
thanks @djhworld 👏 you rock.
